text/0048-synchronizer-redesign.md

```diff
@@ -0,0 +1,170 @@
+
+Based on that prioritization we want to send requests so that:
+
+* We request each vertex from only one peer at a time, but with an aggressive timeout.
+* We don't request too many vertices from the same peer at once. The number will be configurable. A low default — even 1 or 2 — should be fine for a running node, but we should determine experimentally whether a higher value speeds up syncing for a joining node.
+
+Finally, adding vertices to the protocol state as soon as possible is preferable over keeping them in the synchronizer queues: It allows full verification and earlier finality detection. So when syncing, we prefer to download the protocol state bottom-up: the earlier vertices before the later ones. That is why we will add a new way to specify a dependency, so that if we receive unit number 200 from a peer, and we only have unit 100 from that creator yet, we can request unit 101 instead of having to start from unit 199 and moving backwards.
+
+
+## Reference-level explanation
+
+We will move the synchronizer into the `highway_core` module, so it has access to the full content of each vertex, and all internal Highway data.
+
+
+### New dependency type: "Unit with sequence number"
+
+In addition to specifying units by hash, we add a message type
+
+```rust
+UnitBySeqNumber(u64, Hash)
+```
+
+which means we ask for a unit with the specified sequence number by the same sender as the unit with the given hash.
+
+
+### Upon receiving a new vertex
+
+If the vertex was requested from that peer, don't consider the request as pending anymore. (I.e. the sender is eligible to receive more requests.)
+
+If it is a unit with a future timestamp, set a timer for that timestamp and keep it in the queue. When that timer fires, handle it again like a newly received vertex.
+
+If it is missing a dependency or it is a proposal unit that still needs to be validated, keep it in the queue.
+
+Otherwise add it to the Highway protocol state.
+
+Finally, process the queue. (See below.)
+
+
+### Upon completion of block validation
+
+If the block could not be validated because all sending peers timed out, drop the vertex.
+
+If the block is invalid, drop the vertex and everything that depends on it. Ban all the senders of those vertices.
+
+If the block is valid, add the vertex to the protocol state.
+
+Finally, process the queue.
+
+
+### Processing the queue
+
+Iterate over all pending units, sorted by their timestamp, up to the current timestamp, twice:
+
+In the first pass:
+
+* Check whether the unit depends on evidence we don't have, and request it if possible.
```

Comment:
_evidence_ specifically or an type of dependency?

Reply:
Unless you meant to download unit-type dependencies in previous steps and evidence at the last stage since satisfying an evidence request is different from a unit dependency.